src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -0,0 +1,314 @@
+                    || parsedDecorator > MaxArrayRank)
+                {
+#if SYSTEM_PRIVATE_CORELIB
+                    throw new TypeLoadException(); // CLR throws TypeLoadException for invalid decorators
+#else
+                    return null;
+#endif
+                }
+                previousDecorator = parsedDecorator;
+            }
+
+            AssemblyName? assemblyName = null;
+            if (allowFullyQualifiedName && !TryParseAssemblyName(ref assemblyName))
+            {
+#if SYSTEM_PRIVATE_CORELIB
+                // backward compat: throw FileLoadException for non-empty invalid strings
+                if (!_throwOnError && _inputString.TrimStart().StartsWith(","))
+                {
+                    return null;
+                }
+                throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
+#else
+                return null;
+#endif
+            }
+
+            // No matter what was parsed, the full name string is allocated only once.
+            // In case of generic, nested, array, pointer and byref types the full name is allocated
+            // when needed for the first time .
+            string fullName = fullTypeName.ToString();
+
+            TypeName? declaringType = GetDeclaringType(fullName, nestedNameLengths, assemblyName);
+            TypeName result = new(fullName, assemblyName, declaringType: declaringType);
+            if (genericArgs is not null)
+            {
+                result = new(fullName: null, assemblyName, elementOrGenericType: result, declaringType, genericArgs);
+            }
+
+            if (previousDecorator != default) // some decorators were recognized
+            {
+                while (TryParseNextDecorator(ref capturedBeforeProcessing, out int parsedModifier))
+                {
+                    result = new(fullName: null, assemblyName, elementOrGenericType: result, rankOrModifier: (sbyte)parsedModifier);
+                }
+            }
+
+            return result;
+        }
+
+        /// <returns>false means the input was invalid and parsing has failed. Empty input is valid and returns true.</returns>
+        private bool TryParseAssemblyName(ref AssemblyName? assemblyName)
+        {
+            ReadOnlySpan<char> capturedBeforeProcessing = _inputString;
+            if (TryStripFirstCharAndTrailingSpaces(ref _inputString, ','))
+            {
+                if (_inputString.IsEmpty)
+                {
+                    _inputString = capturedBeforeProcessing; // restore the state
+                    return false;
+                }
+
+                // The only delimiter which can terminate an assembly name is ']'.
+                // Otherwise EOL serves as the terminator.
+                int assemblyNameLength = (int)Math.Min((uint)_inputString.IndexOf(']'), (uint)_inputString.Length);
```

Comment:
Is a `']'` character valid in an `AssemblyName` string? If so, this `IndexOf` might split the input string in the wrong place and produce incorrect output.

Reply:
I believe it's not, as the CLR parser rejects escaped `[` here: https://github.com/dotnet/runtime/blob/44e29c2d4e151e35fa4794dbdaab59412ed064cd/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs#L411